src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java

```diff
@@ -148,21 +148,33 @@ public void afterCommand() {
     Map<ModuleExtensionId, ImmutableMap<ModuleExtensionEvalFactors, LockFileModuleExtension>>
         updatedExtensionMap = new HashMap<>();
 
-    // Keep old extensions if they are still valid.
+    // Keep those per factor extension results that are still used according to the static
+    // information given in the extension declaration (dependence on os and arch, reproducibility).
+    // Other information such as transitive .bzl hash and usages hash are *not* checked here.
     for (var entry : oldExtensionInfos.entrySet()) {
       var moduleExtensionId = entry.getKey();
-      var factorToLockedExtension = entry.getValue();
-      ModuleExtensionEvalFactors firstEntryFactors =
-          factorToLockedExtension.keySet().iterator().next();
-      // All entries for a single extension share the same usages digest, so it suffices to check
-      // the first entry.
-      if (shouldKeepExtension(
-          moduleExtensionId,
-          firstEntryFactors,
-          newExtensionInfos.get(moduleExtensionId),
-          depGraphValue)) {
-        updatedExtensionMap.put(moduleExtensionId, factorToLockedExtension);
+      if (!depGraphValue.getExtensionUsagesTable().containsRow(moduleExtensionId)) {
+        // Extensions without any usages are not needed anymore.
+        continue;
+      }
+      var newExtensionInfo = newExtensionInfos.get(moduleExtensionId);
+      if (newExtensionInfo == null) {
+        // No information based on which we could invalidate old entries, keep all of them.
+        updatedExtensionMap.put(moduleExtensionId, entry.getValue());
+        continue;
+      }
+      if (!newExtensionInfo.moduleExtension().shouldLockExtension()) {
+        // Extension has become reproducible and should not be locked anymore.
+        continue;
       }
+      var newFactors = newExtensionInfo.extensionFactors();
+      var perFactorResultsToKeep =
```

Comment:
I think it'd help readability to write this type out. Yes it's very long but still :)